crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -364,6 +397,109 @@ impl Client {
         }
         }
     }
+
+    pub fn class_definitions_stream(
+        self,
+        mut start: BlockNumber,
+        stop_inclusive: BlockNumber,
+        declared_class_counts_stream: impl futures::Stream<Item = anyhow::Result<usize>>,
+    ) -> impl futures::Stream<Item = anyhow::Result<PeerData<Class>>> {
+        async_stream::try_stream! {
+        pin_mut!(declared_class_counts_stream);
+
+        if start <= stop_inclusive {
+            // Loop which refreshes peer set once we exhaust it.
+            'outer: loop {
+                let peers = self
+                    .get_update_peers_with_sync_capability(protocol::Classes::NAME)
+                    .await;
+
+                // Attempt each peer.
+                'next_peer: for peer in peers {
+                    let limit = stop_inclusive.get() - start.get() + 1;
+
+                    let request = ClassesRequest {
+                        iteration: Iteration {
+                            start: start.get().into(),
+                            direction: Direction::Forward,
+                            limit,
+                            step: 1.into(),
+                        },
+                    };
+
+                    let mut responses =
+                        match self.inner.send_classes_sync_request(peer, request).await {
+                            Ok(x) => x,
+                            Err(error) => {
+                                // Failed to establish connection, try next peer.
+                                tracing::debug!(%peer, reason=%error, "Classes request failed");
+                                continue 'next_peer;
+                            }
+                        };
+
+                    // Get number of declared classes for this block
+                    let mut current = declared_class_counts_stream.next().await
+                        .ok_or_else(|| anyhow::anyhow!("Declared class counts stream terminated prematurely at block {start}"))??;
+
+                    while let Some(contract_diff) = responses.next().await {
+                        match contract_diff {
+                            ClassesResponse::Class(p2p_proto::class::Class::Cairo0 {
+                                class,
+                                domain: _,
+                                class_hash,
+                            }) => {
+                                yield PeerData::new(
+                                    peer,
+                                    Class::Cairo {
+                                        block_number: start,
+                                        hash: ClassHash(class_hash.0),
+                                        definition: CairoDefinition::try_from_dto(class)?.0,
+                                    },
```

Comment:
Did you consider just accumulating all classes for a given block before passing them on?

Reply:
Not really, because classes can get pretty big.

Reply:
That's fair; so long as that initial sql query scales well enough it should be fine.